File: code/hashing/horner.cpp
// Horner's method: fold a string into one index, base R = 31.
int hash(const string& s, int M) {
    long h = 0;
    for (char c : s)
        h = (R * h + c) % M;     // reduce each step; no overflow
    return (int)h;
}

File: code/hashing/horner.rs
// Horner's method: fold a string into one index, base R = 31.
fn hash(s: &str, m: i64) -> i64 {
    let mut h: i64 = 0;
    for c in s.chars() {
        h = (R * h + c as i64) % m;   // reduce each step
    }
    h
}
